datasync/chaindatafetcher/kafka/consumer.go

```diff
@@ -101,20 +158,127 @@ func (c *Consumer) Cleanup(s sarama.ConsumerGroupSession) error {
+// Assumption:
+//  1. it is guaranteed that the order of segments is correct.
+//  2. the inserted messages may be duplicated.
+//
+// We can consider the following cases.
+// case1. new segment with index 0 is inserted into newly created segment slice.
+// case2. new consecutive segment is inserted into the right position.
+// case3. duplicated segment is ignored.
+// case4. new sparse segment shouldn't be given, so return an error.
+func insertSegment(segment *Segment, buffer [][]*Segment) ([][]*Segment, error) {
+	for idx, segments := range buffer {
+		length := len(segments)
+		if length > 0 && segments[0].key == segment.key {
+			// there is a missing segment which should not exist.
+			if segment.index > uint64(length) {
+				return buffer, errors.New(missingSegmentErrorMsg)
+			}
+
+			// the segment is already inserted to buffer.
+			if segment.index < uint64(length) {
+				logger.Warn("the message is duplicated", "segment", segment)
+				return buffer, nil
+			}
+
+			// insert the segment to the buffer.
+			buffer[idx] = append(segments, segment)
+			return buffer, nil
+		}
+	}
+
+	if segment.index == 0 {
+		// create a segment slice and append it.
+		buffer = append(buffer, []*Segment{segment})
+	} else {
+		// the segment may be already handled.
+		logger.Warn("the message may be inserted already. drop the segment", "segment", segment)
+	}
+	return buffer, nil
+}
+
+// handleBufferedMessages handles all consecutive complete messages in the buffer.
+func (c *Consumer) handleBufferedMessages(buffer [][]*Segment) ([][]*Segment, error) {
+	for len(buffer) > 0 {
+		// if any message exists in the buffer
+		oldestMsg, firstSegment, buffered := buffer[0], buffer[0][0], len(buffer[0])
+		if uint64(buffered) != firstSegment.total {
+			// not ready for assembling messages
+			return buffer, nil
+		}
+
+		// ready for assembling message
+		var msgBuffer []byte
+		for _, segment := range oldestMsg {
+			msgBuffer = append(msgBuffer, segment.value...)
+		}
+		msg := &sarama.ConsumerMessage{
+			Key:   []byte(firstSegment.key),
+			Value: msgBuffer,
+		}
+
+		f, ok := c.handlers[firstSegment.orig.Topic]
+		if !ok {
+			return buffer, fmt.Errorf("%v: %v", noHandlerErrorMsg, msg.Topic)
+		}
+
+		if err := f(msg); err != nil {
+			return buffer, err
+		}
+
+		buffer = buffer[1:]
+	}
+
+	return buffer, nil
+}
+
+// updateOffset updates offset after handling messages.
+// The offset should be marked for the oldest message (which is not read) in the given buffer.
+// If there is no segment in the buffer, the last consumed message offset should be marked.
+func (c *Consumer) updateOffset(buffer [][]*Segment, lastMsg *sarama.ConsumerMessage, session ConsumerGroupSession) error {
+	if len(buffer) > 0 {
+		if len(buffer[0]) <= 0 {
+			return errors.New(emptySegmentErrorMsg)
+		}
+
+		oldestMsg := buffer[0][0].orig
+		session.MarkOffset(oldestMsg.Topic, oldestMsg.Partition, oldestMsg.Offset, "")
+	} else {
+		session.MarkMessage(lastMsg, "")
```

Comment:
The offset marked in `if` branch indicates **restarting point**, but the offset marked in `else` branch indicates **(restarting point - 1)**.
I think it could have consistency if similar logic of `updateOffset` is executed inside of `handleBufferedMessages`.

Reply:
Oh, I noticed that `MarkMessage` uses lastMsg's offset+1. Nevermind my comment.